### PR TITLE
KAFKA-6093; log.close() should not prevent log from being accessed

### PR DESCRIPTION
**Pull request by lindong28**

### CODE REVIEW

**Comment by lindong28** [comment]
@junrao Can you take a look at this fix? This patch picks up the fix that has already been made in trunk.

File: core/src/main/scala/kafka/log/Log.scala [hunk 1]
```diff
@@ -151,7 +151,9 @@ class Log(@volatile var dir: File,
 
   /* A lock that guards all modifications to the log */
   private val lock = new Object
-  @volatile private var isClosed = false
+  // The memory mapped buffer for index files of this log will be closed for index files of this log will be closed with either delete() or closeHandlers()
```
Comment (tedyu) [inline core/src/main/scala/kafka/log/Log.scala]: 'this log will be closed' is repeated

Comment (lindong28) [reply core/src/main/scala/kafka/log/Log.scala]: Thank you. It is fixed now.

File: core/src/test/scala/unit/kafka/log/LogTest.scala [hunk 2]
```diff
@@ -2137,6 +2137,7 @@ class LogTest {
     for (_ <- 0 until 100)
       log.appendAsLeader(createRecords, leaderEpoch = 0)
 
+    log.close()
```
Comment (ijuma) [inline core/src/test/scala/unit/kafka/log/LogTest.scala]: Do we also want to have a test where we `delete` without calling `close`?

**Comment by junrao** [comment]
@lindong28 : Thanks for the patch. LogManager.asyncDelete() calls removedLog.close() in 1.0, but not in trunk. Does that mean we are leaking file handlers during delete in trunk?

File: core/src/test/scala/unit/kafka/log/LogTest.scala (hunk 2 above)
Comment (lindong28) [reply core/src/test/scala/unit/kafka/log/LogTest.scala]: It seems that we already have a couple of tests in LogTest where `log.delete()` is called without calling `close`. So I guess it is fine. But I can revert the change here and add a new test to call delete() after close(). What do you think?

**Comment by lindong28** [comment]
@junrao Thanks for the review. I think the code in trunk will not leak file handlers because LogManager.deleteLogs() calls Log.delete(). And I verified that Log.delete() will close handlers of all index files and the log segment file. So it should be OK.

File: core/src/test/scala/unit/kafka/log/LogTest.scala (hunk 2 above)
Comment (ijuma) [reply core/src/test/scala/unit/kafka/log/LogTest.scala]: The only thing is that the other tests don't assert anything after the `delete` call. We could add asserts to the other tests or add a new test to make sure we're not losing coverage for that case.

Comment (lindong28) [reply core/src/test/scala/unit/kafka/log/LogTest.scala]: Sure. I have updated the patch to add a new test.

**Comment by junrao** [comment]
@lindong28 : Thanks for the explanation. LGTM. I will let Ismael make another pass and then merge it.

**Comment by ijuma** [comment]
Retest this please

**Comment by ijuma** [comment]
Seems fine to me, I didn't look in great detail, but Jun has and he is more familiar with the JBOD changes. Merged to 1.0. @lindong28, please close the PR.

**Comment by lindong28** [comment]
Thank you all for your review!

I will add that test to trunk in a future patch.

**Pull request closed without merging**